File: apple-watch/Sources/Services/WeatherService.swift
import Foundation

// ── WeatherService.swift ──────────────────────────────────────────────────────
// Calls the same fetchWeather Cloud Function used by the phone app.
// Callable function REST format: POST with {"data": {"lat": x, "lon": y}}

struct WeatherResponse: Codable {
    let result: WeatherResult
    struct WeatherResult: Codable {
        let windMps: Double
        let windDeg: Double
        let gustMps: Double
        let airTempC: Double
        let pressureHpa: Double
        let conditions: String?
    }
}

enum WeatherService {
    // Fill in your Firebase project ID — same as EXPO_PUBLIC_FIREBASE_PROJECT_ID
    static let projectId = "YOUR_FIREBASE_PROJECT_ID"
    static let region = "us-central1"

    private static var functionURL: URL {
        URL(string: "https://\(region)-\(projectId).cloudfunctions.net/fetchWeather")!
    }

    static func fetch(lat: Double, lon: Double) async -> WatchWeatherSummary? {
        var request = URLRequest(url: functionURL, timeoutInterval: 5)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: [
            "data": ["lat": lat, "lon": lon]
        ])

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            let resp = try JSONDecoder().decode(WeatherResponse.self, from: data)
            let r = resp.result
            let sample = WatchWeatherSample(tSec: 0, windMps: r.windMps, windDeg: r.windDeg,
                                            gustMps: r.gustMps, airTempC: r.airTempC,
                                            pressureHpa: r.pressureHpa, conditions: r.conditions)
            return WatchWeatherSummary(start: sample, samples: [sample])
        } catch {
            return nil
        }
    }
}

File: wear/app/src/main/java/app/imuatrak/wear/services/WeatherService.kt
package app.imuatrak.wear.services

import app.imuatrak.wear.BuildConfig
import app.imuatrak.wear.models.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import kotlinx.serialization.json.*
import java.net.HttpURLConnection
import java.net.URL

object WeatherService {
    // Injected at build time from EXPO_PUBLIC_FIREBASE_PROJECT_ID (see
    // app/build.gradle.kts). Blank disables the fetch entirely.
    private const val REGION = "us-central1"
    private val projectId get() = BuildConfig.FIREBASE_PROJECT_ID

    suspend fun fetch(lat: Double, lon: Double): WatchWeatherSummary? {
        if (projectId.isBlank()) return null
        return withTimeoutOrNull(5_000L) {
            withContext(Dispatchers.IO) {
                try {
                    val url = URL("https://$REGION-$projectId.cloudfunctions.net/fetchWeather")
                    val conn = url.openConnection() as HttpURLConnection
                    conn.requestMethod = "POST"
                    conn.setRequestProperty("Content-Type", "application/json")
                    conn.doOutput = true
                    conn.outputStream.write("""{"data":{"lat":$lat,"lon":$lon}}""".toByteArray())

                    val resp = Json.parseToJsonElement(conn.inputStream.bufferedReader().readText())
                    val result = resp.jsonObject["result"]?.jsonObject ?: return@withContext null

                    fun d(key: String) = result[key]?.jsonPrimitive?.double ?: 0.0
                    val sample = WatchWeatherSample(0.0, d("windMps"), d("windDeg"),
                        d("gustMps"), d("airTempC"), d("pressureHpa"),
                        result["conditions"]?.jsonPrimitive?.contentOrNull)
                    WatchWeatherSummary(sample, listOf(sample))
                } catch (_: Exception) { null }
            }
        }
    }
}
